reconcile/utils/terrascript_client.py

```diff
@@ -4073,17 +4093,98 @@ def populate_tf_resource_secrets_manager(self, resource, namespace_info):
 
         self.add_resources(account, tf_resources)
 
-    def populate_tf_resource_asg(self, resource, namespace_info):
+    def _get_commit_sha(self, repo_info: dict) -> str:
+        url = repo_info['url']
+        ref = repo_info['ref']
+        pattern = r'^[0-9a-f]{40}$'
+        # get commit_sha from ref
+        if re.match(pattern, ref):
+            return ref
+        # get commit_sha from branch
+        elif 'github' in url:
+            github = self.init_github()
+            repo_name = url.rstrip("/").replace('https://github.com/', '')
+            repo = github.get_repo(repo_name)
+            commit = repo.get_commit(sha=ref)
+            return commit.sha
+        elif 'gitlab' in url:
+            raise ValueError("dose not support gitlab repo for now")
+
+        return ''
+
+    def _get_asg_image_id(self, image: dict,
+                          existing_secrets: dict, output_prefix: str,
+                          account: str, region: str) -> Tuple[str, str]:
+        """
+        AMI ID comes form queries result or AWS Api filter result.
+        AMI needs to be built by ci account and shared with target account.
+        AMI needs to be taged with a tag_name and
+        its value need to be the commit sha comes from upstream repo.
+        In case of there is a new commit while the packer job is still running,
+        use the AMI form existing asg.
+        """
+        image_id = image.get('id')
+        if image_id:
+            return image_id, ''
+
+        image_repo = image['repo']
+        commit_sha = self._get_commit_sha(image_repo)
+        tag_name = image_repo['tag_name']
+
+        # Get the most recent AMI id
+        aws_accounts = queries.get_ci_aws_accounts()
+        ci_account = aws_accounts[0]
+        ci_account_name = ci_account['name']
+        if account != ci_account_name:
+            aws_accounts.append(self.accounts[account])
+        aws = AWSApi(1, aws_accounts,
+                     settings=self.settings,
+                     init_users=False)
+        owner = [ci_account['uid']]
+        filters: Sequence[FilterTypeDef] = [{
+            'Name': 'tag:' + tag_name,
+            'Values': [commit_sha]
+        }]
+        image_id = aws.get_image_id(ci_account_name, region, owner,
+                                    filters)
+        if not image_id:
```

Comment:
i saw that you imported queries to this file. we generally don't want to involve our graphql schema with any of our utils. this is why queries lives in reconcile and not in reconcile.utils.

that got me thinking - why do we even need queries? to get the ci account. why do we even need the ci account? to check if the image exists there. why do we even care if it exists there if it's not shared with the account we are in the context of?

we probably don't.
this means you can remove this whole part and do something like:
Check if the ami has been shared with target account. if not - log a warning and try to return the commit sha from existing_secrets.

this may simplify things significantly?

Reply:
assuming we can get `image_id` without looking into the ci account...

Reply:
we've discussed it this morning. the `image_id` is obtained from the source account by tag. these tags are not available in the shared account. so we decided to automate the sharing and copying of tags!
see #2216